src/shell/shell_cmds.cpp

```diff
@@ -465,26 +472,231 @@ static std::string to_search_pattern(const char *arg)
 	char buffer[CROSS_LEN];
 	pattern = ExpandDot(pattern.c_str(), buffer, sizeof(buffer));
 
-	// When there's no wildcard and target is a directory then search files
-	// inside the directory.
-	const char *p = pattern.c_str();
-	if (!strrchr(p, '*') && !strrchr(p, '?')) {
-		uint16_t attr = 0;
-		if (DOS_GetFileAttr(p, &attr) && (attr & DOS_ATTR_DIRECTORY))
-			pattern += "\\*.*";
+	return pattern;
+}
+
+uint32_t byte_count,file_count,dir_count;
+Bitu p_count;
+std::vector<std::string> dirs, adirs;
+
+static size_t GetPauseCount() {
+	Bit8u rows=real_readb(BIOSMEM_SEG,BIOSMEM_NB_ROWS);
+	assert(rows>0&&rows<100);
+	if (rows<3) rows=3;
+	return rows;
+}
+
+// With DIR /P, check if it is time to pause ("press any key")
+static bool dirPaused(DOS_Shell * shell, Bitu w_size, bool optP, bool optW, Bit32u &w_count) {
+	p_count+=w_count?1:w_size;								// Increase count
+	if (optP && !(p_count%(GetPauseCount()*w_size))) {		// Time to pause
+		if (w_count&&optW&&w_count%5) {shell->WriteOut("\n");w_count=0;}
+		shell->WriteOut(MSG_Get("SHELL_CMD_PAUSE"));		// Show pause message
+		Bit8u c;Bit16u n=1;
+		DOS_ReadFile(STDIN,&c,&n);							// Do pause (wait for key)
+		if (c==3) {shell->WriteOut("^C\r\n");return false;}	// Break if Ctrl+C
+		if (c==0) DOS_ReadFile(STDIN,&c,&n);				// Read extended key
+		shell->WriteOut_NoParsing("\n");
 	}
+	return true;
+}
 
-	// If no extension, list all files.
-	// This makes patterns like foo* work.
-	if (!strrchr(pattern.c_str(), '.'))
-		pattern += ".*";
+Bit32u wdummy=0;
+static bool doDir(DOS_Shell * shell, char * args, DOS_DTA dta, char * numformat, Bitu w_size, bool optW, bool optZ, bool optS, bool optP, bool optB, bool optA, bool optAD, bool optAminusD, bool optAS, bool optAminusS, bool optAH, bool optAminusH, bool optAR, bool optAminusR, bool optAA, bool optAminusA, bool optO, bool optOG, bool optON, bool optOD, bool optOE, bool optOS, bool reverseSort) {
```

Comment:
@dreamer In order to implement the /S option (list all subdirectories) for DIR, the function has to be recursive. As a result, the bulk of DIR code will be moved to this new function instead of the main function. So the point is that the updates to the DIR command cannot be small. In order to make the first PR small and focused, perhaps we just start with the VER command instead?